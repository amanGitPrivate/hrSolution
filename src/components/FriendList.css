:local(.friendList) {
  padding-left: 0;
  margin-bottom: 0;
}

:local(.friendListContentWrapper) {
  padding-left: 0;
  margin-bottom: 0;
}


:local(.friendListPagination) {
  display: flex;
  justify-content: center;
  padding: 5px;
}

:local(.noFriendsMessage) {
  text-align: center;
  background-color: whitesmoke;
}

