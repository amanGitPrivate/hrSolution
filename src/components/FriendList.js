import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './FriendList.css';
import FriendListItem from './FriendListItem';
import AddFriendPagination from './AddFriendPagination';

class FriendList extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      page: 1,
      itemsPerPage: 2
    };
    this.goToPrevPage = this.goToPrevPage.bind(this);
    this.goToNextPage = this.goToNextPage.bind(this);
    this.isFirstPage = this.isFirstPage.bind(this);
    this.isLastPage = this.isLastPage.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.friends.length !== this.props.friends.length) {
      const totalPages = Math.ceil(
        this.props.friends.length / this.state.itemsPerPage
      );
      if (this.state.page > totalPages || this.state.page === 0) {
        this.setState({ page: totalPages }, () => {
          this.calculatePageItems(this.state.page);
        });
      } else {
        this.calculatePageItems(this.state.page);
      }
    }
  }

  goToPrevPage() {
    this.setState({ page: this.state.page - 1 });
  }

  goToNextPage() {
    this.setState({ page: this.state.page + 1 });
  }

  isFirstPage() {
    return this.state.page === 1;
  }

  isLastPage() {
    const { page, itemsPerPage } = this.state;
    const { friends } = this.props;
    return friends.length <= page * itemsPerPage;
  }

  calculatePageItems(page) {
    const { itemsPerPage } = this.state;
    const { friends } = this.props;

    const itemIndex = page * itemsPerPage - itemsPerPage;
    return [...friends].splice(itemIndex, itemsPerPage);
  }

  render() {
    const pageItems = this.calculatePageItems(this.state.page);
    const { friends } = this.props;
    const { page, itemsPerPage } = this.state;
    return (
      <div className={styles.friendListContentWrapper}>
        <ul className={styles.friendList}>
          {pageItems.length > 0 ? (
            pageItems.map((friend, index) => {
              return (
                <FriendListItem
                  key={index}
                  id={index}
                  {...this.state}
                  {...friend}
                  {...this.props.actions}
                />
              );
            })
          ) : (
            <div className={styles.noFriendsMessage}>
              All items for this page have been deleted
            </div>
          )}
        </ul>
        {friends.length > 2 && (
          <div className={styles.friendListPagination}>
            <AddFriendPagination
              goToPrevPage={this.goToPrevPage}
              goToNextPage={this.goToNextPage}
              isFirstPage={this.isFirstPage}
              isLastPage={this.isLastPage}
              pageNumber={page}
              friends={friends}
              itemsPerPage={itemsPerPage}
            />
          </div>
        )}
      </div>
    );
  }
}

FriendList.propTypes = {
  friends: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired
};

export default FriendList;
