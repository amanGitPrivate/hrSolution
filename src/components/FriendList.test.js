import React from 'react';
import FriendList from './FriendList';
import renderer from 'react-test-renderer';

const friends = [
  {
    name: 'Theodore Roosevelt',
    starred: true
  },
  {
    name: 'Abraham Lincoln',
    starred: false
  },
  {
    name: 'George Washington',
    starred: false
  }
];

const starFriend = jest.fn();
const deleteFriend = jest.fn();

test('FriendList renders correctly', () => {
  const component = renderer.create(
    <FriendList
      friends={friends}
      actions={{ starFriend: starFriend, deleteFriend: deleteFriend }}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FriendList goToPrevPage click test', () => {
  const component = renderer.create(
    <FriendList
      friends={friends}
      actions={{ starFriend: starFriend, deleteFriend: deleteFriend }}
    />
  );
  const instance = component.root.instance;
  instance.state.page = 2;
  instance.goToPrevPage();
  console.assert(instance.state.page === 1);
});
