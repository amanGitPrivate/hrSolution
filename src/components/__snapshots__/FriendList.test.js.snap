exports[`test FriendList renders correctly 1`] = `
<div
  className={undefined}>
  <ul
    className={undefined}>
    <li
      className={undefined}>
      <div
        className={undefined}>
        <div>
          <span>
            Theodore Roosevelt
          </span>
          <i
            className="fa fa-undefined" />
        </div>
        <div>
          <small>
            xx friends in common
          </small>
        </div>
      </div>
      <div
        className={undefined}>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-star" />
        </button>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-trash" />
        </button>
      </div>
    </li>
    <li
      className={undefined}>
      <div
        className={undefined}>
        <div>
          <span>
            Abraham Lincoln
          </span>
          <i
            className="fa fa-undefined" />
        </div>
        <div>
          <small>
            xx friends in common
          </small>
        </div>
      </div>
      <div
        className={undefined}>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-star-o" />
        </button>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-trash" />
        </button>
      </div>
    </li>
  </ul>
  <div
    className={undefined}>
    <div
      className="">
      <button
        className="btn btn-default"
        disabled={true}
        onClick={[Function]}>
        <i
          className="fa fa-caret-left" />
      </button>
      <label
        className={undefined}>
        1
         of 
        2
      </label>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}>
        <i
          className="fa fa-caret-right" />
      </button>
    </div>
  </div>
</div>
`;
